### DISCUSSION TITLE
TypeScript: setting type of service based on machine declaration

### DISCUSSION THREAD

**Discussion by pke** [statelyai/xstate#2653] [Q&A] [answered]
In my tests I want to interpret a machine `beforeEach` 
So I just need to define a `service` variable, but I want to set its types correctly. 

Currently I am using a hack to infer the types of the interpreted machine, by just assigning the result of `interpret` as you can see below.

```ts
describe("hangman state chart", function() {
  // FIXME: just using interpret here to infer correct type of service
  let service = interpret(hangman)

  beforeEach(function() {
    service = interpret(hangman).start()
  })
})
```

However I'd like to just define the type of `service` properly using the `Interpreter` type but since its a generic and requires types for context and event I don't know how to derive those, without having to export `Context` and `Event` from the machine module (just for the typings)

Something like: `let service: Interpreter<typeof machine>`

**Comment by davidkpiano** [comment, marked as answer]
You can do `InterpreterFrom<typeof machine>`:

```ts
import { InterpreterFrom } from 'xstate';

let service: InterpreterFrom<typeof machine>;
```

**Reply by pke** [reply]
Haha I was close. Thanks
